wall/wall.c

```diff
@@ -0,0 +1,125 @@
+/*
+ * wall.c - write to all users
+ *
+ * Copyright (C) 2024: Luiz Antônio Rangel (takusuman)
+ *
+ * SPDX-Licence-Identifier: Zlib
+ *
+ *	from Unix 32V /usr/src/cmd/wall.c
+ *	December 19th, 1978.
+ *
+ * Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
+ *
+ * SPDX-Licence-Identifier: Caldera
+ */
+
+#include <errno.h>
+#include <limits.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strmenta.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <utmp.h>
+#include <utmpx.h>
+#define MAXNAMLEN _POSIX_LOGIN_NAME_MAX
+
+char	*mesg = "",
+	who[MAXNAMLEN] = "???",
+	sterm[32] = "";
+int	msize = 0;
+void main(int argc, char *argv[]);
+void sendmes(char *tty);
+
+void main(int argc, char *argv[]) {
+	int i = 0;
+	char c = '\0',
+	     smesg[3000] = "";
+	struct utmpx *utmp;
+	struct passwd *pw;
+	FILE *f = NULL;
+
+	if ((utmp = getutxent()) == NULL) {
+		fprintf(stderr, "failed to open utmpx database: %s\n",
+			strerror(errno));
+		exit(1);
+	}
+
+	f = stdin;
+	if(argc >= 2) {
+		/* take message from unix file instead of standard input */
+		if((f = fopen(argv[1], "r")) == NULL) {
+			fprintf(stderr, "Cannot open %s\n", argv[1]);
+			exit(1);
+		}
+	}
+	
+	for (; ((i = getc(f)) != EOF); ) smesg[msize++] = i;
+	fclose(f);
+	mesg = ssafe(smesg);
+	
+	pw = getpwuid(geteuid());
+	if (pw) {
+		for (i = 0; c = pw->pw_name[i]; i++) who[i]=c;
+		who[i] = '\0'; /* sender initials */
+	}
+
+	/* Get the current terminal */
+	strncpy(sterm, ttyname(fileno(stderr)), sizeof(sterm));
```

Comment:
It is good to set the last byte of the buffer to `\0` after any usage of `strncpy` in order to avoid buffer overruns

Reply:
Done.